### PR TITLE
add ToString expression

### PR DESCRIPTION
**Pull request by hasancemcerit**
PR based on original suggestion found here. #241

### CODE REVIEW

File: src/Dommel/SqlExpression.cs [hunk 1]
```diff
@@ -419,6 +421,22 @@ protected virtual object VisitContainsExpression(MethodCallExpression expression
             return $"lower({column}) like lower({paramName})";
         }
 
+        /// <summary>
+        /// Processes ToString expression to CAST columns into Char before comparison.
```
Comment (henkmollema) [inline src/Dommel/SqlExpression.cs:425]:
```suggestion
        /// Processes ToString expression to CAST columns into CHAR before comparison.
```

File: src/Dommel/SqlExpression.cs [hunk 2]
```diff
@@ -419,6 +421,22 @@ protected virtual object VisitContainsExpression(MethodCallExpression expression
...
+        /// </summary>
+        /// <param name="expression"></param>
```
Comment (henkmollema) [inline src/Dommel/SqlExpression.cs:427]:
```suggestion
```
remove this line

File: src/Dommel/SqlExpression.cs [hunk 3]
```diff
@@ -419,6 +421,22 @@ protected virtual object VisitContainsExpression(MethodCallExpression expression
...
+        /// <returns></returns>
```
Comment (henkmollema) [inline src/Dommel/SqlExpression.cs:428]:
```suggestion
```
remove this line

File: src/Dommel/SqlExpression.cs [hunk 4]
```diff
@@ -419,6 +421,22 @@ protected virtual object VisitContainsExpression(MethodCallExpression expression
...
+        protected virtual object VisitToStringExpression(MethodCallExpression expression)
+        {
+            var column = VisitExpression(expression.Object);
+            if (expression.Arguments.Count >= 1)
+            {
+                throw new ArgumentException("ToString-expression should not contain any argument.", nameof(expression));
```
Comment (henkmollema) [inline src/Dommel/SqlExpression.cs:434]: Can you add a test case which verifies this behavior? E.g.
```cs
var ex = Assert.Throws<ArgumentException>(() => _sqlExpression.Where(p => p.CategoryId.ToString("n2") == "1"));
Assert.Equal("ToString-expression should not contain any argument.", ex.Message);
```

File: test/Dommel.Tests/SqlExpressions/SqlExpressionTests.cs [hunk 5]
```diff
@@ -12,5 +12,14 @@ public void ToString_ReturnsSql()
             Assert.Equal(" where ([Name] = @p1)", sql);
             Assert.Equal(sql, _sqlExpression.ToString());
         }
+
+        [Fact]
+        public void ToStringVisitation_ReturnsSql()
+        {
+            var _sqlExpression = new SqlExpression<Product>(new SqlServerSqlBuilder());
```
Comment (henkmollema) [inline test/Dommel.Tests/SqlExpressions/SqlExpressionTests.cs:19]: Local variables should not be prefixed. Rename `_sqlExpression` to `sqlExpression`.

**Review by henkmollema — CHANGES_REQUESTED** [review]
Looks good, apart from some minor changes and an extra test case.

File: test/Dommel.Tests/SqlExpressions/SqlExpressionTests.cs (hunk 5 above)
Comment (hasancemcerit) [reply test/Dommel.Tests/SqlExpressions/SqlExpressionTests.cs]: Took the liberty to apply that to your prev. commit 😄

File: src/Dommel/SqlExpression.cs (hunk 4 above)
Comment (hasancemcerit) [reply src/Dommel/SqlExpression.cs:433]: done.

**Review by henkmollema — COMMENTED** [review]
Great! I think an integration test should be added as well to verify this behavior on the different database systems. I'm no sure if the `CAST(XXX as CHAR)` syntax is supported on all platforms. Can you look into that or do you want me to do that?

**Comment by hasancemcerit** [comment]
Good point.  Here's the matrix.
MS SQL and MySQL/MariaDB, SQLLite supports CHAR

image

image

image
⚠️  SQLLite adds e notation to large numbers, but I guess that is not a concern.

PostgreSQL does `NOT` support CHAR, default char(1), but supports VARCHAR
image

I see an option to add `Cast` method to DB specific builder interfaces, like you have on limit or pagination, or create an abstract class to allow for future type conversion differences.  

@henkmollema your call...

**Comment by henkmollema** [comment]
Sorry for the late reaction. Let's merge this for now and perhaps add a SQL builder interface method in the future to configure this behavior.

**Pull request merged**